src/v/coproc/offset_storage_utils.cc

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright 2020 Vectorized, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "coproc/offset_storage_utils.h"
+
+#include "bytes/iobuf.h"
+#include "coproc/logger.h"
+#include "model/fundamental.h"
+#include "reflection/adl.h"
+#include "vlog.h"
+
+#include <seastar/core/coroutine.hh>
+
+#include <optional>
+
+namespace coproc {
+
+void offsets_map_to_iobuf(
+  iobuf& buf, const ntp_context::offset_tracker& offsets) {
+    reflection::serialize(buf, offsets.size());
+    for (const auto& e : offsets) {
+        reflection::serialize(
+          buf, e.first, e.second.last_read, e.second.last_acked);
+    }
+}
+
+ntp_context::offset_tracker rebuild_offsets_map(iobuf_parser& p) {
+    ntp_context::offset_tracker offsets;
+    auto size = reflection::adl<ntp_context::offset_tracker::size_type>{}.from(
+      p);
+    for (auto i = 0; i < size; ++i) {
+        auto id = reflection::adl<script_id>{}.from(p);
+        auto last_read = reflection::adl<model::offset>{}.from(p);
+        auto last_acked = reflection::adl<model::offset>{}.from(p);
+        offsets.emplace(
+          id,
+          ntp_context::offset_pair{
+            .last_read = last_read, .last_acked = last_acked});
+    }
+    return offsets;
+}
+
+ss::future<ntp_context_cache> recover_offsets(
+  storage::snapshot_manager& snap, storage::log_manager& log_mgr) {
+    using size_type = ntp_context::offset_tracker::size_type;
+    ntp_context_cache recovered;
+    auto optional_snap_reader = co_await snap.open_snapshot();
+    if (!optional_snap_reader) {
+        co_return recovered;
+    }
+    storage::snapshot_reader& reader = *optional_snap_reader;
+    vlog(coproclog.info, "Recovering coprocessor offsets....");
+    /// Read the metadata which contains the total number of elements
+    iobuf metadata = co_await reader.read_metadata();
+    const size_type n_elements
```

Comment:
is this `n_trackers` ?

Reply:
Its not the number of scripts registered, its the number of unique input ntps registered